src/common/src/util/value_encoding/mod.rs

```diff
@@ -31,6 +35,190 @@ use error::ValueEncodingError;
+                    .for_each(|m| self.offsets.put_u16_le(m as u16));
+            }
+            n if n <= u32::MAX as usize => {
+                self.flag |= 0b11;
+                maybe_offset
+                    .into_iter()
+                    .for_each(|m| self.offsets.put_u32_le(m as u32));
+            }
+            _ => unreachable!("encoding length exceeds u32"),
+        }
+    }
+
+    pub fn encode(
+        &mut self,
+        datum_refs: impl Iterator<Item = impl ToDatumRef>,
+        column_ids: Vec<ColumnId>,
+    ) {
+        assert!(
+            self.buf.is_empty() && self.non_null_column_ids.is_empty(),
+            "should not encode one RowEncoding object multiple times."
+        );
+        let mut maybe_offset = vec![];
+        for (datum, &id) in datum_refs.zip_eq(column_ids.iter()) {
+            if let Some(v) = datum.to_datum_ref() {
+                maybe_offset.push(self.buf.len());
+                self.non_null_column_ids.push(id);
+                serialize_scalar(v, &mut self.buf);
+            }
+        }
+        let max_offset = *maybe_offset
+            .last()
+            .expect("should encode at least one column");
+        self.set_width(self.non_null_column_ids.len());
+        self.set_big(maybe_offset, max_offset);
+    }
+
+    pub fn serialize(&self) -> Vec<u8> {
+        let mut row_bytes = vec![];
+        row_bytes.put_u8(self.flag);
+        serialize_width(self.non_null_datum_nums, &mut row_bytes);
+        for id in &self.non_null_column_ids {
+            row_bytes.put_i32_le(id.get_id());
+        }
+        row_bytes.extend(self.offsets.iter());
+        row_bytes.extend(self.buf.iter());
+
+        row_bytes
+    }
+}
+
+pub fn serialize_row_column_aware(column_ids: Vec<ColumnId>, row: impl Row) -> Vec<u8> {
+    let encoding = encode_datums(column_ids, row.iter());
+    encoding.serialize()
+}
+
+pub fn encode_datums(
+    column_ids: Vec<ColumnId>,
+    datum_refs: impl Iterator<Item = impl ToDatumRef>,
+) -> RowEncoding {
+    let mut encoding = RowEncoding::new();
+    // encoding.update_non_null_datum_nums(column_ids);
+    encoding.encode(datum_refs, column_ids);
+    encoding
+}
+
+pub fn decode(
+    need_columns: &[ColumnId],
+    schema: &[DataType],
+    mut encoded_bytes: &[u8],
+) -> Vec<Datum> {
+    let flag = encoded_bytes.get_u8();
+    let nums_bytes = match flag & 0b1100 {
+        0b0100 => 1,
+        0b1000 => 2,
+        0b1100 => 4,
+        _ => unreachable!("flag's WW bits corrupted"),
+    };
+    let offset_bytes = match flag & 0b11 {
+        0b01 => 1,
+        0b10 => 2,
+        0b11 => 4,
+        _ => unreachable!("flag's BB bits corrupted"),
+    };
+    let non_null_datum_nums = deserialize_width(nums_bytes, &mut encoded_bytes);
+    let id_to_index = (0..non_null_datum_nums)
+        .into_iter()
+        .fold(HashMap::new(), |mut map, i| {
+            map.insert(ColumnId::new(encoded_bytes.get_i32_le()), i);
+            map
+        });
```

Comment:
Same here. This should be done only once, instead of per call. May refer to the existing `Deserializer` or `Serializer` structures.